### PR TITLE
Checkout order

### PR DESCRIPTION
**Pull request by rajneeshsharma9**

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    email = params[:order][:receiver_email]
+    if @order.update_receiver_email!(email)
+      redirect_to edit_order_path, success: 'Enter billing address'
+    else
+      redirect_to edit_order_path, danger: @order.halted_because
+    end
+  end
+
+  private def update_address
+    address = permitted_address_params
```
Comment (sawangupta92) [inline app/controllers/orders_controller.rb:51]: why do we need this variable?

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      update_email
+    elsif @order.email?
+      update_address
+    elsif @order.address?
+      confirm
+    end
+  end
+
+  private def update_email
+    email = params[:order][:receiver_email]
```
Comment (sawangupta92) [inline app/controllers/orders_controller.rb:42]: why do we need this variable?

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  def update_cart
+    if @order.update_cart(permitted_order_params)
+      redirect_to cart_path, success: success_message(permitted_order_params[:task])
+    else
+      redirect_to cart_path, danger: error_message
+    end
+  end
+
+  def index
+    @orders = current_user.orders.where(workflow_state: 'completed')
```
Comment (sawangupta92) [inline app/controllers/orders_controller.rb:26]: does workflow provides scope for the same?

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  private def set_current_user
+    @order.current_user = current_user
+  end
+
+  private def permitted_order_params
+    params.permit(:deal_id, :task)
+  end
+
+  private def permitted_address_params
+    params[:order][:address].permit!.to_h
```
Comment (sawangupta92) [inline app/controllers/orders_controller.rb:76]: let's make a habit to never use permit!, always permit specific params regardless of the number of params

File: app/models/concerns/line_items_handler.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
+module LineItemsHandler
+
+  private def create_line_item
+    current_item = line_items.find_by(deal_id: @deal.id)
```
Comment (sawangupta92) [inline app/models/concerns/line_items_handler.rb:4]: L:4 and L:5 can be merged

File: app/models/concerns/line_items_handler.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+module LineItemsHandler
+
+  private def create_line_item
```
Comment (sawangupta92) [inline app/models/concerns/line_items_handler.rb:3]: rename

File: app/models/concerns/line_items_handler.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    @line_item.decrement(:quantity).save
+  end
+
+  private def set_line_item
+    # no need to check if line_item found or not as add_deal handles it
+    @line_item = line_items.find_by(deal_id: @params[:deal_id])
+  end
+
+  private def set_deal
+    @deal = Deal.find_by(id: @params[:deal_id])
```
Comment (sawangupta92) [inline app/models/concerns/line_items_handler.rb:25]: find for active deals only

File: app/models/deal.rb [hunk 8]
```diff
@@ -51,6 +52,22 @@ def published?
     published_at.present?
   end
 
+  def expired?
+    expire_at < Time.current
+  end
+
+  def quantity_sold
+    orders.includes(:line_items).where(workflow_state: 'completed').sum('line_items.quantity')
```
Comment (sawangupta92) [inline app/models/deal.rb:60]: do we need includes? or can we use join?

Comment (sawangupta92) [inline app/models/deal.rb:60]: .where(workflow_state: 'completed')

File: app/models/order.rb [hunk 9]
```diff
@@ -1,20 +1,53 @@
...
   validates :amount, numericality: { greater_than_or_equal_to: MINIMUM_ALLOWED_AMOUNT, less_than_or_equal_to: MAXIMUM_ALLOWED_AMOUNT }, allow_nil: true
+  validates :line_items, length: { maximum: 1, message: ': Only one deal allowed in cart at a time. Please remove it from cart first.' }
   validates :receiver_email, format: {
     with:    EMAIL_REGEXP,
     message: :invalid_email
   }, allow_blank: true
 
+  def update_cart(params)
+    @params = params # need params in before callbacks
+    run_callbacks :update_cart do
```
Comment (sawangupta92) [inline app/models/order.rb:34]: can you confirm if this all happens in a transaction

File: app/models/user.rb [hunk 10]
```diff
@@ -25,4 +26,9 @@ class User < ApplicationRecord
   }, allow_blank: true
   validates :verification_token, uniqueness: { case_sensitive: false }, allow_nil: true
   validates :password, length: { in: PASSWORD_VALIDATION_RANGE }, allow_blank: true
+
+  def purchased_deal_quantity(deal_id)
+    orders.includes(:line_items).where(workflow_state: 'completed').sum { |order| order.line_items.where(deal_id: deal_id).sum(&:quantity) }
```
Comment (sawangupta92) [inline app/models/user.rb:31]: can this be optimized?

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: app/models/deal.rb (hunk 8 above)
Comment (rajneeshsharma9) [reply app/models/deal.rb:60]: we can use both.

**Pull request merged**